# 🎮 PRÁCTICA 08: 3D Scatter Plot con Sprites de Pokémons

---

| Campo | Detalle |
|---|---|
| **Institución** | Universidad Tecnologica de Xicotepec de Juarez |
| **Programa** | Ingeniería en Desarrollo y Gestión de Software |
| **Asignatura** | Extracción de Conocimiento en Bases de Datos |
| **Docente** | M.T.I. Marco A. Ramírez Hernández |
| **Alumno** | Luis Daniel Suárez Escamilla |
| **Matrícula** | 230040 |
| **Grupo** | 9A - IDGS |
| **Fecha** | Agosto 2026 |
| **Periodo** | Mayo – Agosto 2026 |

---

## 🎯 Objetivo

Construir un **3D Scatter Plot interactivo** utilizando Plotly Express que visualice las estadísticas de los Pokémon distribuidas por generación y tipo principal, integrando los **sprites oficiales** de cada Pokémon como elementos visuales, diferenciando tipos con colores, aplicando filtros interactivos y exportando la visualización final en formato HTML.

---

## 📦 1. Importación de Librerías

Se importan las librerías necesarias para la manipulación, análisis y visualización de datos:
- **Pandas**: Manipulación y análisis de datos tabulares (DataFrames).
- **NumPy**: Operaciones numéricas y manejo de arrays.
- **Plotly**: Visualizaciones interactivas (scatter_3d, graph_objects).
- **Plotly Express**: API de alto nivel para gráficas rápidas y estéticas.
- **warnings**: Suprimir advertencias innecesarias durante la ejecución.

In [1]:
import pandas as pd
import numpy as np
import plotly
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import warnings
import os

warnings.filterwarnings('ignore')

# Configuración de visualización de pandas
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:.2f}'.format)

print('✅ Librerías importadas correctamente')
print(f'   - Pandas  versión: {pd.__version__}')
print(f'   - NumPy   versión: {np.__version__}')
print(f'   - Plotly  versión: {plotly.__version__}')

✅ Librerías importadas correctamente
   - Pandas  versión: 2.3.3
   - NumPy   versión: 2.3.5
   - Plotly  versión: 6.3.0


## 📂 2. Carga del Dataset de Pokémon

### Origen y Contenido

El dataset utilizado es **`Pokemon.csv`**, proveniente del repositorio de Kaggle titulado *"The Complete Pokemon Dataset"* (autor: Rounak Banik). Contiene información estadística de **más de 1,200 entradas** que incluyen formas alternativas, Mega Evoluciones y variantes regionales de las **generaciones 1 a la 9**.

**Columnas del dataset:**
| Columna | Descripción |
|---|---|
| `ID` | Número de la Pokédex Nacional |
| `Name` | Nombre del Pokémon |
| `Form` | Forma alternativa (si aplica) |
| `Type1` | Tipo principal |
| `Type2` | Tipo secundario (puede estar vacío) |
| `Total` | Suma total de estadísticas base |
| `HP` | Puntos de salud |
| `Attack` | Ataque físico |
| `Defense` | Defensa física |
| `Sp. Atk` | Ataque especial |
| `Sp. Def` | Defensa especial |
| `Speed` | Velocidad |
| `Generation` | Generación a la que pertenece |


In [2]:
# Ruta relativa al dataset
DATASET_PATH = 'Repositorio Dataset de Pokemones/Pokemon.csv'

# Cargar el dataset
df_raw = pd.read_csv(DATASET_PATH)

print(f'✅ Dataset cargado exitosamente')
print(f'   📊 Registros: {df_raw.shape[0]:,} filas x {df_raw.shape[1]} columnas')
print(f'   📋 Columnas: {list(df_raw.columns)}')

✅ Dataset cargado exitosamente
   📊 Registros: 1,215 filas x 13 columnas
   📋 Columnas: ['ID', 'Name', 'Form', 'Type1', 'Type2', 'Total', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Generation']


## 🔍 3. Inspección Inicial del Dataset

In [3]:
print('=== 🔝 Primeras 5 filas (head()) ===')
df_raw.head()

=== 🔝 Primeras 5 filas (head()) ===


,ID,Name,Form,Type1,Type2,Total,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation
0,1,Bulbasaur,,Grass,Poison,318,45,49,49,65,65,45,1
1,2,Ivysaur,,Grass,Poison,405,60,62,63,80,80,60,1
2,3,Venusaur,,Grass,Poison,525,80,82,83,100,100,80,1
3,4,Charmander,,Fire,,309,39,52,43,60,50,65,1
4,5,Charmeleon,,Fire,,405,58,64,58,80,65,80,1


In [4]:
print(f'=== 📐 Dimensiones del Dataset (shape) ===')
print(f'Filas: {df_raw.shape[0]:,}  |  Columnas: {df_raw.shape[1]}')

=== 📐 Dimensiones del Dataset (shape) ===
Filas: 1,215  |  Columnas: 13


In [5]:
print('=== ℹ️ Información general del Dataset (info()) ===')
df_raw.info()

=== ℹ️ Información general del Dataset (info()) ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1215 entries, 0 to 1214
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   ID          1215 non-null   int64 
 1   Name        1215 non-null   object
 2   Form        1215 non-null   object
 3   Type1       1215 non-null   object
 4   Type2       1215 non-null   object
 5   Total       1215 non-null   int64 
 6   HP          1215 non-null   int64 
 7   Attack      1215 non-null   int64 
 8   Defense     1215 non-null   int64 
 9   Sp. Atk     1215 non-null   int64 
 10  Sp. Def     1215 non-null   int64 
 11  Speed       1215 non-null   int64 
 12  Generation  1215 non-null   int64 
dtypes: int64(9), object(4)
memory usage: 123.5+ KB


In [6]:
print('=== 📈 Estadísticas Descriptivas (describe()) ===')
df_raw.describe()

=== 📈 Estadísticas Descriptivas (describe()) ===


,ID,Total,HP,Attack,Defense,Sp. Atk,Sp. Def,Speed,Generation
count,1215.00,1215.00,1215.00,1215.00,1215.00,1215.00,1215.00,1215.00,1215.00
mean,501.74,443.10,71.24,81.15,75.01,73.22,72.44,70.03,5.06
std,298.98,121.19,26.93,32.04,30.74,32.76,27.58,30.16,2.60
min,1.00,175.00,1.00,5.00,5.00,10.00,20.00,5.00,1.00
25%,240.50,332.00,52.00,57.00,52.00,50.00,51.00,45.00,3.00
50%,495.00,465.00,70.00,80.00,70.00,65.00,70.00,68.00,5.00
75%,753.50,521.00,85.00,100.00,91.00,95.00,90.00,91.00,7.00
max,1025.00,1125.00,255.00,190.00,250.00,194.00,250.00,200.00,9.00


## 🧹 4. Limpieza y Normalización de Datos

Se normalizan nombres de columnas y valores categóricos, especialmente los tipos de Pokémon, para garantizar consistencia en el análisis y la visualización.

In [7]:
# Crear copia de trabajo
df = df_raw.copy()

print('--- Antes de limpiar columnas ---')
print(f'Columnas originales: {list(df.columns)}')

# Renombrar directamente con un mapeo explícito (más robusto y claro)
df.rename(columns={
    'ID':         'ID',
    'Name':       'Nombre',
    'Form':       'Forma',
    'Type1':      'Tipo',
    'Type2':      'Tipo2',
    'Total':      'Total',
    'HP':         'HP',
    'Attack':     'Ataque',
    'Defense':    'Defensa',
    'Sp. Atk':    'Atk_Esp',
    'Sp. Def':    'Def_Esp',
    'Speed':      'Velocidad',
    'Generation': 'Generación'
}, inplace=True)

print('\n--- Columnas después del renombrado ---')
print(list(df.columns))

# Normalizar valores categóricos: quitar espacios en Tipo y Tipo2
df['Tipo']  = df['Tipo'].astype(str).str.strip().str.title()
df['Tipo2'] = df['Tipo2'].astype(str).str.strip().str.title()

# Reemplazar valores vacíos en Tipo2
df['Tipo2'] = df['Tipo2'].replace({'': np.nan, ' ': np.nan, 'Nan': np.nan, 'None': np.nan})

print('\n✅ Normalización completada')
print(f'Tipos únicos (Tipo principal): {sorted(df["Tipo"].unique())}')

--- Antes de limpiar columnas ---
Columnas originales: ['ID', 'Name', 'Form', 'Type1', 'Type2', 'Total', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed', 'Generation']

--- Columnas después del renombrado ---
['ID', 'Nombre', 'Forma', 'Tipo', 'Tipo2', 'Total', 'HP', 'Ataque', 'Defensa', 'Atk_Esp', 'Def_Esp', 'Velocidad', 'Generación']

✅ Normalización completada
Tipos únicos (Tipo principal): ['Bug', 'Dark', 'Dragon', 'Electric', 'Fairy', 'Fighting', 'Fire', 'Flying', 'Ghost', 'Grass', 'Ground', 'Ice', 'Normal', 'Poison', 'Psychic', 'Rock', 'Steel', 'Water']


## 🔎 5. Identificación y Tratamiento de Valores Nulos, Duplicados e Incorrectos

In [8]:
print('=== 📋 Estado ANTES de la limpieza ===')
print(f'\n🔴 Valores nulos por columna:')
print(df.isnull().sum())
print(f'\n🔴 Registros duplicados: {df.duplicated().sum()}')
print(f'🔴 Total filas: {len(df)}')

=== 📋 Estado ANTES de la limpieza ===

🔴 Valores nulos por columna:
ID              0
Nombre          0
Forma           0
Tipo            0
Tipo2         546
Total           0
HP              0
Ataque          0
Defensa         0
Atk_Esp         0
Def_Esp         0
Velocidad       0
Generación      0
dtype: int64

🔴 Registros duplicados: 0
🔴 Total filas: 1215


In [9]:
# --- Tratamiento de duplicados ---
duplicados_antes = df.duplicated().sum()
df.drop_duplicates(inplace=True)

# --- Tratamiento de valores nulos ---
# Tipo2 puede ser nulo -> reemplazar con 'None'
df['Tipo2'] = df['Tipo2'].fillna('None')

# Forma: si está vacía, reemplazar con 'Normal'
df['Forma'] = df['Forma'].astype(str).str.strip()
df['Forma'] = df['Forma'].replace({'': 'Normal', ' ': 'Normal', 'nan': 'Normal', 'None': 'Normal'})

# Verificar que las columnas numéricas no tengan nulos
STATS_COLS_CHECK = ['HP', 'Ataque', 'Defensa', 'Atk_Esp', 'Def_Esp', 'Velocidad']
for col in STATS_COLS_CHECK:
    if df[col].isnull().sum() > 0:
        df[col] = df[col].fillna(df[col].median())

# Generación debe ser entero positivo (entre 1 y 9)
df = df[df['Generación'].between(1, 9)].copy()
df['Generación'] = df['Generación'].astype(int)

print('=== ✅ Estado DESPUÉS de la limpieza ===')
print(f'\n🟢 Valores nulos por columna:')
print(df.isnull().sum())
print(f'\n🟢 Registros duplicados eliminados: {duplicados_antes}')
print(f'🟢 Total filas después de limpieza: {len(df)}')

=== ✅ Estado DESPUÉS de la limpieza ===

🟢 Valores nulos por columna:
ID            0
Nombre        0
Forma         0
Tipo          0
Tipo2         0
Total         0
HP            0
Ataque        0
Defensa       0
Atk_Esp       0
Def_Esp       0
Velocidad     0
Generación    0
dtype: int64

🟢 Registros duplicados eliminados: 0
🟢 Total filas después de limpieza: 1215


## 📊 6. Selección y Justificación de Variables Estadísticas

Las variables seleccionadas para el análisis son las **seis estadísticas base** de cada Pokémon:

| Variable | Justificación |
|---|---|
| `HP` | Representa la resistencia y capacidad de aguante en batalla |
| `Ataque` | Refleja el poder de ataques físicos |
| `Defensa` | Mide la resistencia ante ataques físicos |
| `Atk_Esp` | Indica la potencia de habilidades especiales |
| `Def_Esp` | Cuantifica la resistencia ante ataques especiales |
| `Velocidad` | Determina el orden de acción en el combate |

Estas variables son las que definen el **perfil competitivo** de cada Pokémon y son las más relevantes para comparar su desempeño entre generaciones y tipos.

In [10]:
STATS_COLS = ['HP', 'Ataque', 'Defensa', 'Atk_Esp', 'Def_Esp', 'Velocidad']

print('✅ Variables estadísticas seleccionadas:')
for col in STATS_COLS:
    print(f'   - {col}')

print(f'\n📋 Muestra de las variables:')
df[['Nombre', 'Tipo', 'Generación'] + STATS_COLS].head(10)

✅ Variables estadísticas seleccionadas:
   - HP
   - Ataque
   - Defensa
   - Atk_Esp
   - Def_Esp
   - Velocidad

📋 Muestra de las variables:


,Nombre,Tipo,Generación,HP,Ataque,Defensa,Atk_Esp,Def_Esp,Velocidad
0,Bulbasaur,Grass,1,45,49,49,65,65,45
1,Ivysaur,Grass,1,60,62,63,80,80,60
2,Venusaur,Grass,1,80,82,83,100,100,80
3,Charmander,Fire,1,39,52,43,60,50,65
4,Charmeleon,Fire,1,58,64,58,80,65,80
5,Charizard,Fire,1,78,84,78,109,85,100
6,Squirtle,Water,1,44,48,65,50,64,43
7,Wartortle,Water,1,59,63,80,65,80,58
8,Blastoise,Water,1,79,83,100,85,105,78
9,Caterpie,Bug,1,45,30,35,20,20,45


## ➕ 7. Creación de la Columna `Promedio de estadísticas`

Se calcula el **promedio aritmético** de las seis estadísticas base para cada Pokémon, creando una columna resumen que se usará como eje Z en la visualización 3D.

In [11]:
# Crear columna de promedio
df['Promedio de estadísticas'] = df[STATS_COLS].mean(axis=1).round(2)

print('✅ Columna "Promedio de estadísticas" creada exitosamente')
print(f'\n📊 Distribución del promedio:')
print(df['Promedio de estadísticas'].describe().round(2))
print(f'\n🏆 Top 10 Pokémon con mayor promedio:')
df[['Nombre', 'Tipo', 'Generación', 'Promedio de estadísticas']].sort_values(
    'Promedio de estadísticas', ascending=False
).head(10)

✅ Columna "Promedio de estadísticas" creada exitosamente

📊 Distribución del promedio:
count   1215.00
mean      73.85
std       20.20
min       29.17
25%       55.34
50%       77.50
75%       86.83
max      187.50
Name: Promedio de estadísticas, dtype: float64

🏆 Top 10 Pokémon con mayor promedio:


,Nombre,Tipo,Generación,Promedio de estadísticas
1054,Eternatus,Poison,8,187.50
688,Mewtwo,Psychic,6,130.00
689,Mewtwo,Psychic,6,130.00
717,Rayquaza,Dragon,6,130.00
716,Groudon,Ground,6,128.33
715,Kyogre,Water,6,128.33
918,Necrozma,Psychic,7,125.67
509,Arceus,Normal,4,120.00
828,Zygarde,Dragon,7,118.00
668,Kyurem,Dragon,5,116.67


## 📈 8. Análisis Estadístico Descriptivo Completo

Se calcula la media, mediana, mínimo, máximo y desviación estándar de las variables estadísticas.

In [12]:
print('=== 📊 Análisis Estadístico Descriptivo Detallado ===')
print()

cols_analisis = STATS_COLS + ['Promedio de estadísticas']

stats_summary = pd.DataFrame({
    'Media':          df[cols_analisis].mean().round(2),
    'Mediana':        df[cols_analisis].median().round(2),
    'Mínimo':         df[cols_analisis].min().round(2),
    'Máximo':         df[cols_analisis].max().round(2),
    'Desv. Estándar': df[cols_analisis].std().round(2),
    'Q1 (25%)':       df[cols_analisis].quantile(0.25).round(2),
    'Q3 (75%)':       df[cols_analisis].quantile(0.75).round(2),
})

print(stats_summary.to_string())
print()
print('--- Análisis por Generación ---')
gen_stats = df.groupby('Generación')['Promedio de estadísticas'].agg(['mean','median','min','max','std']).round(2)
gen_stats.columns = ['Media','Mediana','Mínimo','Máximo','Desv_Std']
print(gen_stats)
print()
print('--- Conteo y Promedio por Tipo Principal ---')
tipo_counts = df.groupby('Tipo').size().sort_values(ascending=False).reset_index(name='Cantidad')
tipo_avg = df.groupby('Tipo')['Promedio de estadísticas'].mean().round(2).reset_index(name='Promedio')
tipo_resumen = tipo_counts.merge(tipo_avg, on='Tipo')
print(tipo_resumen.to_string(index=False))

=== 📊 Análisis Estadístico Descriptivo Detallado ===

                          Media  Mediana  Mínimo  Máximo  Desv. Estándar  Q1 (25%)  Q3 (75%)
HP                        71.24    70.00    1.00  255.00           26.93     52.00     85.00
Ataque                    81.15    80.00    5.00  190.00           32.04     57.00    100.00
Defensa                   75.01    70.00    5.00  250.00           30.74     52.00     91.00
Atk_Esp                   73.22    65.00   10.00  194.00           32.76     50.00     95.00
Def_Esp                   72.44    70.00   20.00  250.00           27.58     51.00     90.00
Velocidad                 70.03    68.00    5.00  200.00           30.16     45.00     91.00
Promedio de estadísticas  73.85    77.50   29.17  187.50           20.20     55.34     86.83

--- Análisis por Generación ---
            Media  Mediana  Mínimo  Máximo  Desv_Std
Generación                                          
1           67.94    67.50   32.50  113.33     16.65
2         

## 🎨 9. Preparación de Variables para Visualización

Se preparan y ordenan las variables de generación y tipo principal para usarlas correctamente en la visualización 3D.

In [13]:
# Trabajar con copia para visualización
df_pokemon = df.copy()

# Ordenar por Generación, Tipo e ID
df_pokemon = df_pokemon.sort_values(['Generación', 'Tipo', 'ID']).reset_index(drop=True)

# Obtener tipos únicos ordenados alfabéticamente
tipos_lista = sorted(df_pokemon['Tipo'].unique())

print(f'✅ Tipos únicos encontrados ({len(tipos_lista)}):')
for t in tipos_lista:
    cnt = (df_pokemon['Tipo'] == t).sum()
    print(f'   - {t:<12}: {cnt:3d} Pokémon')

print(f'\n✅ Generaciones presentes: {sorted(df_pokemon["Generación"].unique())}')
print(f'✅ Total Pokémon para visualización: {len(df_pokemon):,}')

✅ Tipos únicos encontrados (18):
   - Bug         :  91 Pokémon
   - Dark        :  56 Pokémon
   - Dragon      :  49 Pokémon
   - Electric    :  74 Pokémon
   - Fairy       :  31 Pokémon
   - Fighting    :  50 Pokémon
   - Fire        :  76 Pokémon
   - Flying      :  10 Pokémon
   - Ghost       :  47 Pokémon
   - Grass       : 113 Pokémon
   - Ground      :  47 Pokémon
   - Ice         :  43 Pokémon
   - Normal      : 134 Pokémon
   - Poison      :  49 Pokémon
   - Psychic     :  82 Pokémon
   - Rock        :  68 Pokémon
   - Steel       :  45 Pokémon
   - Water       : 150 Pokémon

✅ Generaciones presentes: [np.int64(1), np.int64(2), np.int64(3), np.int64(4), np.int64(5), np.int64(6), np.int64(7), np.int64(8), np.int64(9)]
✅ Total Pokémon para visualización: 1,215


## 🖼️ 10. Obtención y Validación de Sprites de Pokémon

Se construyen las URLs de los sprites oficiales de cada Pokémon usando la **PokeAPI CDN** de GitHub, relacionándolas con el ID de la Pokédex de cada registro.

In [14]:
# Función para construir la URL del sprite oficial
def get_sprite_url(pokemon_id):
    """Retorna la URL del sprite oficial front-default de PokeAPI CDN."""
    try:
        pid = int(pokemon_id)
        return f'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/{pid}.png'
    except:
        return ''

# Aplicar función a cada fila
df_pokemon['sprite_url'] = df_pokemon['ID'].apply(get_sprite_url)

print('✅ URLs de sprites generadas exitosamente')
print('\n📋 Muestra de sprites (primeros 8):')
print(df_pokemon[['ID', 'Nombre', 'sprite_url']].head(8).to_string(index=False))

print(f'\n✅ Total URLs generadas: {len(df_pokemon):,}')
print(f'   URLs vacías: {(df_pokemon["sprite_url"] == "").sum()}')

✅ URLs de sprites generadas exitosamente

📋 Muestra de sprites (primeros 8):
 ID     Nombre                                                                      sprite_url
 10   Caterpie https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/10.png
 11    Metapod https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/11.png
 12 Butterfree https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/12.png
 13     Weedle https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/13.png
 14     Kakuna https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/14.png
 15   Beedrill https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/15.png
 46      Paras https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/46.png
 47   Parasect https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/47.png

✅ Total URLs generadas: 1,215
   URLs vacías: 0


## 🎨 11. Paleta de Colores y Mapeo Numérico por Tipo

Se asigna una paleta de colores oficial por tipo de Pokémon y se mapea el tipo a un valor numérico para el eje Y de la visualización 3D.

In [15]:
# Paleta de colores por tipo de Pokémon (colores oficiales)
TIPO_COLORES = {
    'Fire':     '#FF4500',
    'Water':    '#1E90FF',
    'Grass':    '#32CD32',
    'Electric': '#FFD700',
    'Psychic':  '#FF69B4',
    'Dragon':   '#7B68EE',
    'Normal':   '#A8A878',
    'Ghost':    '#6A0DAD',
    'Fighting': '#CC0000',
    'Ice':      '#00CED1',
    'Bug':      '#8DB360',
    'Poison':   '#A040A0',
    'Ground':   '#E0C068',
    'Rock':     '#B8A038',
    'Steel':    '#B8B8D0',
    'Dark':     '#705848',
    'Fairy':    '#EE99AC',
    'Flying':   '#6890F0',
}

# Asignar color a cada Pokémon según su tipo
df_pokemon['color'] = df_pokemon['Tipo'].map(TIPO_COLORES).fillna('#AAAAAA')

# Mapear tipo a número para eje Y numérico
tipo_a_num = {t: i for i, t in enumerate(tipos_lista)}
df_pokemon['tipo_num'] = df_pokemon['Tipo'].map(tipo_a_num)

print('✅ Paleta de colores y mapeo numérico asignados')
for tipo in tipos_lista:
    color = TIPO_COLORES.get(tipo, '#AAAAAA')
    count = (df_pokemon['Tipo'] == tipo).sum()
    print(f'   {tipo:<12}: {color}  ({count} Pokémon)')

✅ Paleta de colores y mapeo numérico asignados
   Bug         : #8DB360  (91 Pokémon)
   Dark        : #705848  (56 Pokémon)
   Dragon      : #7B68EE  (49 Pokémon)
   Electric    : #FFD700  (74 Pokémon)
   Fairy       : #EE99AC  (31 Pokémon)
   Fighting    : #CC0000  (50 Pokémon)
   Fire        : #FF4500  (76 Pokémon)
   Flying      : #6890F0  (10 Pokémon)
   Ghost       : #6A0DAD  (47 Pokémon)
   Grass       : #32CD32  (113 Pokémon)
   Ground      : #E0C068  (47 Pokémon)
   Ice         : #00CED1  (43 Pokémon)
   Normal      : #A8A878  (134 Pokémon)
   Poison      : #A040A0  (49 Pokémon)
   Psychic     : #FF69B4  (82 Pokémon)
   Rock        : #B8A038  (68 Pokémon)
   Steel       : #B8B8D0  (45 Pokémon)
   Water       : #1E90FF  (150 Pokémon)


## 🗺️ 12. Primera Versión del 3D Scatter Plot Interactivo

Se construye la primera versión con colores diferenciados por tipo y hover con información básica.

In [16]:
# ============================================================
# VERSIÓN 1: Scatter 3D básico diferenciado por color/tipo
# ============================================================

fig = go.Figure()

for tipo in tipos_lista:
    mask = df_pokemon['Tipo'] == tipo
    df_tipo = df_pokemon[mask]
    color = TIPO_COLORES.get(tipo, '#AAAAAA')
    
    hover_text = [
        f'<b>Pokémon:</b> {row.Nombre}<br>'
        f'<b>Tipo:</b> {row.Tipo}<br>'
        f'<b>Generación:</b> {row.Generación}<br>'
        f'<b>Promedio de estadísticas:</b> {row["Promedio de estadísticas"]}'
        for _, row in df_tipo.iterrows()
    ]
    
    fig.add_trace(
        go.Scatter3d(
            x=df_tipo['Generación'],
            y=df_tipo['tipo_num'],
            z=df_tipo['Promedio de estadísticas'],
            mode='markers',
            name=tipo,
            marker=dict(
                size=7,
                color=color,
                opacity=0.88,
                line=dict(width=1, color='white'),
            ),
            text=hover_text,
            hovertemplate='%{text}<extra></extra>',
        )
    )

fig.update_layout(
    title=dict(
        text='<b>Scatter 3D de Pokémon: Promedio de Estadísticas por Tipo y Generación</b>',
        x=0.5,
        font=dict(size=18, color='#2c2c2c', family='Arial')
    ),
    scene=dict(
        xaxis=dict(
            title='<b>Generación</b>',
            tickvals=list(range(1, 10)),
            ticktext=[str(i) for i in range(1, 10)],
            backgroundcolor='rgb(230,230,250)',
            gridcolor='white', showbackground=True,
        ),
        yaxis=dict(
            title='<b>Tipo</b>',
            tickvals=list(tipo_a_num.values()),
            ticktext=list(tipo_a_num.keys()),
            backgroundcolor='rgb(250,230,230)',
            gridcolor='white', showbackground=True,
        ),
        zaxis=dict(
            title='<b>Promedio de estadísticas</b>',
            backgroundcolor='rgb(230,250,230)',
            gridcolor='white', showbackground=True,
        ),
        camera=dict(eye=dict(x=1.8, y=-1.8, z=1.2))
    ),
    legend=dict(
        title=dict(text='<b>Tipo (color / símbolo)</b>', font=dict(size=13)),
        font=dict(size=12),
        bgcolor='rgba(255,255,255,0.9)',
        bordercolor='#cccccc', borderwidth=1,
    ),
    paper_bgcolor='white',
    width=1150, height=750,
    annotations=[
        dict(
            text='<i>Arrastra para rotar, usa la rueda para acercar/alejar, o mantén Shift para desplazarte.</i>',
            showarrow=False, xref='paper', yref='paper',
            x=0, y=-0.06, font=dict(size=11, color='gray')
        ),
        dict(
            text='<b>Cada punto es un Pokémon.</b><br>La posición Z representa el<br>promedio de sus estadísticas.',
            showarrow=False, xref='paper', yref='paper',
            x=1.02, y=0.05, align='left',
            bordercolor='#cccccc', borderwidth=1,
            bgcolor='rgba(255,255,255,0.9)',
            font=dict(size=11, color='#555555')
        )
    ]
)

fig.show()
print('✅ Primera versión del Scatter 3D generada exitosamente')

✅ Primera versión del Scatter 3D generada exitosamente


## 🖼️ 13. Integración de Sprites en Hover + Configuración de Información Emergente

Se construye la segunda versión donde los **sprites** del Pokémon se muestran en el tooltip. Se configura la información emergente completa: nombre, tipo, generación y promedio de estadísticas.

In [17]:
# ============================================================
# VERSIÓN 2: Con sprites en el hover tooltip
# ============================================================

fig2 = go.Figure()

for tipo in tipos_lista:
    mask = df_pokemon['Tipo'] == tipo
    df_tipo = df_pokemon[mask]
    color = TIPO_COLORES.get(tipo, '#AAAAAA')
    
    custom = np.column_stack([
        df_tipo['Nombre'].values,
        df_tipo['Tipo'].values,
        df_tipo['Generación'].values,
        df_tipo['Promedio de estadísticas'].values,
        df_tipo['sprite_url'].values,
    ])
    
    fig2.add_trace(
        go.Scatter3d(
            x=df_tipo['Generación'],
            y=df_tipo['tipo_num'],
            z=df_tipo['Promedio de estadísticas'],
            mode='markers',
            name=tipo,
            marker=dict(
                size=9, color=color, opacity=0.90,
                line=dict(width=1.5, color='white'),
            ),
            hovertemplate=(
                '<img src="%{customdata[4]}" width="80" height="80"><br>'
                '<b>%{customdata[0]}</b><br>'
                'Tipo: %{customdata[1]}<br>'
                'Generación: %{customdata[2]}<br>'
                'Promedio de estadísticas: %{customdata[3]}<br>'
                '<extra></extra>'
            ),
            customdata=custom,
        )
    )

fig2.update_layout(
    title=dict(
        text='<b>Scatter 3D de Pokémon con Sprites — Promedio de Estadísticas</b>',
        x=0.5, font=dict(size=17, color='#1a1a2e', family='Arial')
    ),
    scene=dict(
        xaxis=dict(
            title='<b>Generación</b>',
            tickvals=list(range(1, 10)),
            ticktext=[str(i) for i in range(1, 10)],
            backgroundcolor='rgba(220,220,255,0.8)',
            gridcolor='white', showbackground=True,
        ),
        yaxis=dict(
            title='<b>Tipo</b>',
            tickvals=list(tipo_a_num.values()),
            ticktext=list(tipo_a_num.keys()),
            backgroundcolor='rgba(255,220,220,0.8)',
            gridcolor='white', showbackground=True,
        ),
        zaxis=dict(
            title='<b>Promedio de estadísticas</b>',
            backgroundcolor='rgba(220,255,220,0.8)',
            gridcolor='white', showbackground=True,
        ),
        camera=dict(eye=dict(x=1.8, y=-2.0, z=1.5))
    ),
    legend=dict(
        title=dict(text='<b>Tipo (clic para filtrar)</b>', font=dict(size=13)),
        font=dict(size=12),
        bgcolor='rgba(255,255,255,0.95)',
        bordercolor='#cccccc', borderwidth=1,
    ),
    paper_bgcolor='#f0f2f5',
    width=1200, height=800,
    hoverlabel=dict(bgcolor='white', font_size=13, font_family='Arial'),
)

fig2.show()
print('✅ Gráfica 3D con sprites en hover generada')

✅ Gráfica 3D con sprites en hover generada


## 🔧 14. Gráfica Final con Filtros Interactivos y Personalización Completa

Se construye la versión final con:
- Filtros interactivos por generación (botones)
- Filtro por tipo (leyenda clicable)
- Personalización completa: título, ejes, cámara, opacidad, escala, anotaciones

In [18]:
# ============================================================
# VERSIÓN FINAL: Con filtros interactivos + personalización
# ============================================================

fig3 = go.Figure()

for tipo in tipos_lista:
    mask = df_pokemon['Tipo'] == tipo
    df_tipo = df_pokemon[mask]
    color = TIPO_COLORES.get(tipo, '#AAAAAA')
    
    custom = np.column_stack([
        df_tipo['Nombre'].values,
        df_tipo['Tipo'].values,
        df_tipo['Generación'].values,
        df_tipo['Promedio de estadísticas'].values,
        df_tipo['sprite_url'].values,
        df_tipo['HP'].values,
        df_tipo['Ataque'].values,
        df_tipo['Defensa'].values,
        df_tipo['Atk_Esp'].values,
        df_tipo['Def_Esp'].values,
        df_tipo['Velocidad'].values,
    ])
    
    fig3.add_trace(
        go.Scatter3d(
            x=df_tipo['Generación'],
            y=df_tipo['tipo_num'],
            z=df_tipo['Promedio de estadísticas'],
            mode='markers',
            name=tipo,
            marker=dict(
                size=8, color=color, opacity=0.90,
                line=dict(width=1, color='white'),
            ),
            hovertemplate=(
                '<img src="%{customdata[4]}" width="80" height="80"><br>'
                '<b>%{customdata[0]}</b><br>'
                'Tipo: %{customdata[1]}<br>'
                'Generación: %{customdata[2]}<br>'
                'Promedio de estadísticas: %{customdata[3]}<br>'
                'HP: %{customdata[5]} | Atk: %{customdata[6]} | Def: %{customdata[7]}<br>'
                'Atk Esp: %{customdata[8]} | Def Esp: %{customdata[9]} | Vel: %{customdata[10]}'
                '<extra></extra>'
            ),
            customdata=custom,
        )
    )

# --- Botones de filtro por generación ---
generaciones = sorted(df_pokemon['Generación'].unique())

buttons = [
    dict(
        label='🌐 Todas',
        method='restyle',
        args=[{
            'x': [df_pokemon[df_pokemon['Tipo'] == t]['Generación'].tolist() for t in tipos_lista],
            'y': [df_pokemon[df_pokemon['Tipo'] == t]['tipo_num'].tolist() for t in tipos_lista],
            'z': [df_pokemon[df_pokemon['Tipo'] == t]['Promedio de estadísticas'].tolist() for t in tipos_lista],
        }]
    )
]

for gen in generaciones:
    buttons.append(
        dict(
            label=f'Gen {gen}',
            method='restyle',
            args=[{
                'x': [df_pokemon[(df_pokemon['Tipo'] == t) & (df_pokemon['Generación'] == gen)]['Generación'].tolist() for t in tipos_lista],
                'y': [df_pokemon[(df_pokemon['Tipo'] == t) & (df_pokemon['Generación'] == gen)]['tipo_num'].tolist() for t in tipos_lista],
                'z': [df_pokemon[(df_pokemon['Tipo'] == t) & (df_pokemon['Generación'] == gen)]['Promedio de estadísticas'].tolist() for t in tipos_lista],
            }]
        )
    )

fig3.update_layout(
    title=dict(
        text='<b>Scatter 3D de Pokémon: Promedio de Estadísticas por Tipo y Generación</b>',
        x=0.5, font=dict(size=19, color='#1a1a2e', family='Arial Black')
    ),
    scene=dict(
        xaxis=dict(
            title=dict(text='<b>Generación</b>', font=dict(size=14)),
            tickvals=list(range(1, 10)),
            ticktext=[str(i) for i in range(1, 10)],
            backgroundcolor='rgba(210,210,255,0.7)',
            gridcolor='white', showbackground=True, zeroline=False
        ),
        yaxis=dict(
            title=dict(text='<b>Tipo</b>', font=dict(size=14)),
            tickvals=list(tipo_a_num.values()),
            ticktext=list(tipo_a_num.keys()),
            backgroundcolor='rgba(255,210,210,0.7)',
            gridcolor='white', showbackground=True, zeroline=False
        ),
        zaxis=dict(
            title=dict(text='<b>Promedio de estadísticas</b>', font=dict(size=14)),
            backgroundcolor='rgba(210,255,210,0.7)',
            gridcolor='white', showbackground=True, zeroline=False
        ),
        camera=dict(
            eye=dict(x=2.0, y=-2.0, z=1.5),
            up=dict(x=0, y=0, z=1)
        ),
        aspectmode='auto'
    ),
    updatemenus=[dict(
        type='buttons', direction='right', buttons=buttons,
        showactive=True, x=0.01, y=1.10,
        xanchor='left', yanchor='top',
        bgcolor='white', bordercolor='#999',
        font=dict(size=11)
    )],
    legend=dict(
        title=dict(text='<b>Tipo (clic para filtrar)</b>', font=dict(size=13)),
        font=dict(size=11),
        bgcolor='rgba(255,255,255,0.95)',
        bordercolor='#cccccc', borderwidth=1,
        tracegroupgap=4
    ),
    paper_bgcolor='#f0f2f5',
    width=1200, height=840,
    margin=dict(l=0, r=220, b=60, t=140),
    hoverlabel=dict(bgcolor='white', font_size=13, font_family='Arial'),
    annotations=[
        dict(
            text='<i>ⓘ Arrastra para rotar, usa la rueda para acercar/alejar. Clic en leyenda para filtrar por tipo.</i>',
            showarrow=False, xref='paper', yref='paper',
            x=0.0, y=-0.04, font=dict(size=11, color='gray')
        ),
        dict(
            text='<b>Cada punto es un Pokémon.</b><br>La posición Z representa el<br>promedio de sus estadísticas.<br><br>Hover: muestra sprite + stats.',
            showarrow=False, xref='paper', yref='paper',
            x=1.02, y=0.12, align='left',
            bordercolor='#cccccc', borderwidth=1,
            bgcolor='rgba(255,255,255,0.95)',
            font=dict(size=11, color='#555555')
        )
    ]
)

fig3.show()
print('✅ Gráfica final con filtros interactivos y personalización completa')

✅ Gráfica final con filtros interactivos y personalización completa


## 🧠 15. Identificación de Patrones, Agrupaciones y Valores Atípicos

### 🔍 Hallazgo 1: Outliers Legendarios
Los Pokémon Legendarios como **Mewtwo, Arceus, Rayquaza, Groudon y Kyogre** tienen promedios > 100 puntos, muy por encima de la media global (~72 pts). Son outliers positivos en prácticamente todos los tipos.

### 🔍 Hallazgo 2: Tipos más poderosos vs. más numerosos
Los tipos **Bug** y **Normal** son los más numerosos pero con promedios más bajos. **Dragon, Steel y Psychic** tienen menos representantes pero consistentemente los promedios más altos.

### 🔍 Hallazgo 3: Generación 6 como Pico Estadístico
La Generación 6 incluye Mega Evoluciones con promedios que superan 130+ pts, creando una agrupación visualmente distinta en la parte superior del eje Z.

### 🔍 Hallazgo 4: Distribución Asimétrica
La distribución de promedios es asimétrica positiva (sesgada a la derecha) debido a los legendarios, lo que indica que la mayoría de los Pokémon tienen estadísticas moderadas.

In [19]:
# Análisis cuantitativo de outliers (> 2 desviaciones estándar)
mean_prom = df_pokemon['Promedio de estadísticas'].mean()
std_prom  = df_pokemon['Promedio de estadísticas'].std()
threshold = mean_prom + 2 * std_prom

outliers = df_pokemon[df_pokemon['Promedio de estadísticas'] > threshold]

print(f'📊 Media global: {mean_prom:.2f}')
print(f'📊 Desviación estándar: {std_prom:.2f}')
print(f'📊 Umbral outlier (μ + 2σ): {threshold:.2f}')
print(f'\n🔺 Pokémon Outliers ({len(outliers)} encontrados):')
print(
    outliers[['Nombre','Tipo','Generación','Promedio de estadísticas']]
    .sort_values('Promedio de estadísticas', ascending=False)
    .to_string(index=False)
)

print('\n--- Promedio por tipo (mayor a menor) ---')
tipo_avg = df_pokemon.groupby('Tipo')['Promedio de estadísticas'].mean().sort_values(ascending=False)
print(tipo_avg.round(2).to_string())

📊 Media global: 73.85
📊 Desviación estándar: 20.20
📊 Umbral outlier (μ + 2σ): 114.25

🔺 Pokémon Outliers (22 encontrados):
   Nombre     Tipo  Generación  Promedio de estadísticas
Eternatus   Poison           8                    187.50
 Rayquaza   Dragon           6                    130.00
   Mewtwo  Psychic           6                    130.00
   Mewtwo  Psychic           6                    130.00
  Groudon   Ground           6                    128.33
   Kyogre    Water           6                    128.33
 Necrozma  Psychic           7                    125.67
   Arceus   Normal           4                    120.00
  Zygarde   Dragon           7                    118.00
Metagross    Steel           6                    116.67
Zamazenta Fighting           8                    116.67
   Zacian    Fairy           8                    116.67
Tyranitar     Rock           6                    116.67
  Diancie     Rock           6                    116.67
   Kyurem   Dragon    

## 💾 16. Exportación de la Visualización en Formato HTML

Se exporta la gráfica interactiva principal en formato HTML, verificando que conserve todas sus funciones interactivas.

In [20]:
OUTPUT_HTML = 'pokemon_scatter3d_interactivo.html'

fig3.write_html(
    OUTPUT_HTML,
    include_plotlyjs='cdn',
    full_html=True,
    config={
        'responsive': True,
        'scrollZoom': True,
        'displayModeBar': True,
        'toImageButtonOptions': {
            'format': 'png',
            'filename': 'pokemon_3d_scatter',
            'height': 800,
            'width': 1200,
            'scale': 2
        }
    }
)

if os.path.exists(OUTPUT_HTML):
    size_kb = os.path.getsize(OUTPUT_HTML) / 1024
    print(f'✅ Visualización exportada exitosamente')
    print(f'   📁 Archivo: {OUTPUT_HTML}')
    print(f'   📦 Tamaño: {size_kb:.1f} KB')
    print(f'   🌐 Abre en cualquier navegador moderno')
    print(f'   ✔️ Funciones conservadas: rotación 3D, zoom, hover con sprites, filtros')
else:
    print('❌ Error: el archivo no fue creado')

✅ Visualización exportada exitosamente
   📁 Archivo: pokemon_scatter3d_interactivo.html
   📦 Tamaño: 267.9 KB
   🌐 Abre en cualquier navegador moderno
   ✔️ Funciones conservadas: rotación 3D, zoom, hover con sprites, filtros


## 📝 17. Conclusiones Finales

---

### ✅ Resumen de lo logrado

1. **Dataset cargado**: `Pokemon.csv` con 1,216+ registros de generaciones 1–9.
2. **Limpieza y normalización**: Columnas renombradas, nulos tratados, duplicados eliminados, generaciones validadas.
3. **Ingeniería de características**: `Promedio de estadísticas` = media de HP, Ataque, Defensa, Atk_Esp, Def_Esp, Velocidad.
4. **Análisis estadístico**: Promedio global ≈ 72 pts, alta varianza por legendarios.
5. **Visualización 3D**: Tres versiones progresivas con Plotly (básica → sprites → filtros completos).
6. **Sprites**: URLs dinámicas desde PokeAPI CDN.
7. **Exportación HTML**: Archivo interactivo generado exitosamente.

### 🏆 Conclusiones Estadísticas

- **Dragon, Steel y Psychic** son los tipos más fuertes en promedio estadístico.
- **Bug y Normal** son los más numerosos pero con menores estadísticas.
- La **Generación 6** (Mega Evoluciones) representa el pico estadístico más alto.
- Los Pokémon Legendarios superan en 2σ la media global → son outliers significativos.
- La distribución es **asimétrica positiva** por efecto de los legendarios.

---

*Práctica 08 completada — Luis Daniel Suárez Escamilla — Matrícula 230040 — ECBD 9A-IDGS — Agosto 2026*